ksql-engine/src/main/java/io/confluent/ksql/function/udtf/Cube.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udtf;
+
+import io.confluent.ksql.util.KsqlConstants;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@UdtfDescription(name = "cube_explode", author = KsqlConstants.CONFLUENT_AUTHOR,
+    description =
+        "Takes as argument a list of columns and outputs all possible combinations of them. "
+            + "It produces 2^d new rows where d is the number of columns given as parameter. "
+            + "Duplicate entries for columns with null value are skipped.")
+public class Cube {
+
+  @Udtf
+  public <T> List<List<T>> cube(final List<T> columns) {
+    if (columns == null) {
+      return Collections.emptyList();
+    }
+
+    List<List<T>> result = new ArrayList<>();
+    createAllCombinations(columns, 0, new ArrayList<>(), result);
+
+    return result;
+  }
+
+
+  private <T> void createAllCombinations(List<T> columns, int pos, List<T> current,
```

Comment:
Here's a fun trick. I was wondering if we could write this iteratively without recursion, and this neat solution came to mind that avoids having 2^n stacks and (personally) I think is actually easier to read as well:
```java
    List<String> input = ImmutableList.of("1", "2", "3");
    int combinations = 1 << input.size();

    List<List<String>> result = new ArrayList<>();
    // bitmask is a binary number where a set bit represents that
    // the value at that index of input should be included - iterate
    // backwards so that we start with a full row instead of an empty
    // one
    for (int bitmask = combinations - 1; bitmask >= 0; bitmask--) {
      List<String> row = new ArrayList<>();
      for (int i = 0; i < input.size(); i++) {
        row.add((bitmask & (1 << i)) == 0 ? null : input.get(i));
      }
      result.add(row);
    }

    System.out.println(result);
```

Reply:
That's really really cool! Thanks Almog! Incorporated your suggestion.